stdlib/public/core/SmallString.swift

```diff
@@ -0,0 +1,846 @@
+    _: Encoding.Type = Encoding.self
+  ) where S.Element == Encoding.CodeUnit {
+#if arch(i386) || arch(arm)
+    return nil // Never form small strings on 32-bit
+#else
+    guard utf16Length <= 15 else { return nil }
+
+    // TODO: transcode
+    guard isASCII else { return nil }
+
+    self.init()
+    var bufferIdx = 0
+    for encodedScalar in Unicode._ParsingIterator(
+      codeUnits: codeUnits.makeIterator(),
+      parser: Encoding.ForwardParser()
+    ) {
+      guard let transcoded = Unicode.UTF8.transcode(
+        encodedScalar, from: Encoding.self
+      ) else {
+        fatalError("Somehow un-transcodable?")
+      }
+      _sanityCheck(transcoded.count <= 4, "how?")
+      guard bufferIdx + transcoded.count <= 15 else { return nil }
+      for i in transcoded.indices {
+        self._uncheckedSetCodeUnit(at: bufferIdx, to: transcoded[i])
+        bufferIdx += 1
+      }
+    }
+    _sanityCheck(self.count == 0, "overwrote count early?")
+    self.count = bufferIdx
+    _invariantCheck()
+
+    // FIXME: support transcoding
+    if !self.isASCII { return nil }
+#endif
+  }
+}
+
+extension _SmallUTF8String {
+#if arch(i386) || arch(arm)
+  @_fixed_layout @_versioned struct UnicodeScalarIterator {
+    @_versioned @_inlineable @inline(__always)
+    func next() -> Unicode.Scalar? { unsupportedOn32bit() }
+  }
+  @_versioned @_inlineable @inline(__always)
+  func makeUnicodeScalarIterator() -> UnicodeScalarIterator {
+    unsupportedOn32bit()
+  }
+#else
+  // FIXME (TODO: JIRA): Just make a real decoding iterator
+  @_fixed_layout
+  @_versioned // FIXME(sil-serialize-all)
+  struct UnicodeScalarIterator {
+    @_versioned // FIXME(sil-serialize-all)
+    var buffer: _SmallUTF16StringBuffer
+    @_versioned // FIXME(sil-serialize-all)
+    var count: Int
+    @_versioned // FIXME(sil-serialize-all)
+    var _offset: Int
+
+    @_inlineable // FIXME(sil-serialize-all)
+    @_versioned // FIXME(sil-serialize-all)
+    init(_ base: _SmallUTF8String) {
+      (self.buffer, self.count) = base.transcoded
+      self._offset = 0
+    }
+
+    @_inlineable // FIXME(sil-serialize-all)
+    @_versioned // FIXME(sil-serialize-all)
+    mutating func next() -> Unicode.Scalar? {
+      if _slowPath(_offset == count) { return nil }
+      let u0 = buffer[_offset]
+      if _fastPath(UTF16._isScalar(u0)) {
+        _offset += 1
+        return Unicode.Scalar(u0)
+      }
+      if UTF16.isLeadSurrogate(u0) && _offset + 1 < count {
+        let u1 = buffer[_offset + 1]
+        if UTF16.isTrailSurrogate(u1) {
+          _offset += 2
+          return UTF16._decodeSurrogates(u0, u1)
+        }
+      }
+      _offset += 1
+      return Unicode.Scalar._replacementCharacter
+    }
+  }
+
+  @_versioned // FIXME(sil-serialize-all)
+  @_inlineable
+  func makeUnicodeScalarIterator() -> UnicodeScalarIterator {
+    return UnicodeScalarIterator(self)
+  }
+#endif // 64-bit
+}
+
+#if arch(i386) || arch(arm)
+#else
+extension _SmallUTF8String {
```

Comment:
I tried both ways. In general, I found it a little simpler to have the primary interfaces with the `#if` at the top and the implementation at the bottom, where it isn't riddled with `#if`. This may change when we can remove many of these.